File: backend/app/schemas/strategy_library.py
"""Pydantic schemas for Strategy Library and Protocol Engine."""
from typing import Optional
from pydantic import BaseModel


# --- Strategy Library Item ---
class StrategyLibraryItemBase(BaseModel):
    domain_key: str
    module_key: Optional[str] = None
    name: str
    category: str
    evidence_level: str
    impact_level: str
    difficulty_level: str
    description: Optional[str] = None
    when_to_use: Optional[str] = None
    contraindications: Optional[str] = None
    active: bool = True


class StrategyLibraryItemResponse(StrategyLibraryItemBase):
    id: str
    created_at: Optional[str] = None


# --- Protocol Step ---
class ProtocolStepBase(BaseModel):
    order_index: int = 0
    title: str
    description: Optional[str] = None
    frequency: Optional[str] = None
    target_metric_key: Optional[str] = None
    xp_reward: float = 0


class ProtocolStepResponse(ProtocolStepBase):
    id: str


# --- Strategy Protocol ---
class StrategyProtocolBase(BaseModel):
    name: str
    cadence: str
    duration_days: Optional[int] = None
    instructions_json: Optional[dict] = None


class StrategyProtocolResponse(StrategyProtocolBase):
    id: str
    strategy_id: str
    created_at: Optional[str] = None
    steps: list[ProtocolStepResponse] = []


class StrategyWithProtocolsResponse(StrategyLibraryItemResponse):
    protocols: list[StrategyProtocolResponse] = []


# --- User Active Protocol ---
class UserActiveProtocolBase(BaseModel):
    protocol_id: str
    notes: Optional[str] = None


class UserActiveProtocolResponse(BaseModel):
    id: str
    protocol_id: str
    started_at: Optional[str] = None
    active: bool
    adherence_score: float
    effectiveness_score: Optional[float] = None
    notes: Optional[str] = None


class ActiveProtocolDetailResponse(UserActiveProtocolResponse):
    strategy_name: str
    protocol_name: str
    domain_key: str
    category: str
    evidence_level: str
    impact_level: str
    difficulty_level: str
    module_key: Optional[str] = None
    steps: list[ProtocolStepResponse] = []


# --- Protocol Checkin ---
class ProtocolCheckinCreate(BaseModel):
    completed_steps_json: Optional[list] = None
    adherence_value: Optional[float] = None
    notes: Optional[str] = None


class ProtocolCheckinResponse(BaseModel):
    id: str
    user_active_protocol_id: str
    checked_at: Optional[str] = None
    adherence_value: Optional[float] = None
    notes: Optional[str] = None


# --- Recommendation ---
class StrategyRecommendation(BaseModel):
    strategy_id: str
    strategy_name: str
    protocol_id: str
    protocol_name: str
    why_recommended: str
    evidence_level: str
    estimated_benefit: str
    estimated_effort: str
    domain_key: str
    module_source: Optional[str] = None
    category: str

File: frontend/src/services/strategyLibrary.ts
import { api } from './api'

export interface StrategyLibraryItem {
  id: string
  domain_key: string
  module_key?: string
  name: string
  category: string
  evidence_level: string
  impact_level: string
  difficulty_level: string
  description?: string
  when_to_use?: string
  contraindications?: string
  active: boolean
  created_at?: string
}

export interface ProtocolStep {
  id: string
  order_index: number
  title: string
  description?: string
  frequency?: string
  target_metric_key?: string
  xp_reward: number
}

export interface StrategyProtocol {
  id: string
  name: string
  cadence: string
  duration_days?: number
  instructions_json?: Record<string, unknown>
  created_at?: string
  steps: ProtocolStep[]
}

export interface StrategyWithProtocols extends StrategyLibraryItem {
  protocols: StrategyProtocol[]
}

export interface StrategyRecommendation {
  strategy_id: string
  strategy_name: string
  protocol_id: string
  protocol_name: string
  why_recommended: string
  evidence_level: string
  estimated_benefit: string
  estimated_effort: string
  domain_key: string
  module_source?: string
  category: string
}

export interface ActiveProtocol {
  id: string
  protocol_id: string
  strategy_id: string
  strategy_name: string
  protocol_name: string
  domain_key: string
  category: string
  evidence_level: string
  impact_level: string
  difficulty_level: string
  module_key?: string
  started_at?: string
  active: boolean
  adherence_score: number
  effectiveness_score?: number
  notes?: string
  steps: ProtocolStep[]
}

export const strategyLibraryService = {
  list: (params?: { domain_key?: string; category?: string }) => {
    const q = new URLSearchParams()
    if (params?.domain_key) q.set('domain_key', params.domain_key)
    if (params?.category) q.set('category', params.category)
    const query = q.toString()
    return api.get<StrategyLibraryItem[]>(`/strategy-library${query ? `?${query}` : ''}`)
  },

  listByDomain: (domainKey: string, category?: string) => {
    const q = category ? `?category=${encodeURIComponent(category)}` : ''
    return api.get<StrategyLibraryItem[]>(`/strategy-library/domains/${domainKey}${q}`)
  },

  getDetail: (strategyId: string) =>
    api.get<StrategyWithProtocols>(`/strategy-library/items/${strategyId}`),

  recommendations: (limit = 10, domainKey?: string) => {
    const params = new URLSearchParams({ limit: String(limit) })
    if (domainKey) params.set('domain_key', domainKey)
    return api.get<StrategyRecommendation[]>(`/strategy-library/recommendations?${params}`)
  },

  activeProtocols: () =>
    api.get<ActiveProtocol[]>(`/strategy-library/protocols/active`),

  activateProtocol: (protocolId: string) =>
    api.post<{ id: string; protocol_id: string; started_at?: string; active: boolean }>(
      '/strategy-library/protocols/activate',
      { protocol_id: protocolId }
    ),

  deactivateProtocol: (userActiveProtocolId: string) =>
    api.post<{ success: boolean }>(
      `/strategy-library/protocols/${userActiveProtocolId}/deactivate`,
      {}
    ),

  checkin: (
    userActiveProtocolId: string,
    body: { completed_steps_json?: unknown[]; adherence_value?: number; notes?: string }
  ) =>
    api.post<{ id: string; checked_at?: string; adherence_value?: number }>(
      `/strategy-library/protocols/${userActiveProtocolId}/checkin`,
      body
    ),

  updateAdherence: (userActiveProtocolId: string, score: number) =>
    api.patch<{ success: boolean }>(
      `/strategy-library/protocols/${userActiveProtocolId}/adherence`,
      { score }
    ),
}
